lib/elixir/src/elixir_clauses.erl

```diff
@@ -40,8 +40,18 @@ guard({'when', Meta, [Left, Right]}, E) ->
   {ELeft, EL}  = guard(Left, E),
   {ERight, ER} = guard(Right, EL),
   {{'when', Meta, [ELeft, ERight]}, ER};
-guard(Other, E) ->
-  elixir_expand:expand(Other, E).
+guard(Guard, E) ->
+  warn_zero_length_guard(Guard, E),
+  elixir_expand:expand(Guard, E).
+
+warn_zero_length_guard({'==', Meta, [{length, _, [{Var, _, _}]}, 0]}, E) ->
+  MessageFmt = "`length(~p) == 0` is a bad practice! Use `~p == []` instead!",
```

Comment:
Any idea for a better message?

I would put a link with more information, but do you have any link explaining why not to use `length(x) == 0`?

Reply:
A good error message says what went wrong, why and how to fix it. So what is missing on this error message is exactly why `length(...) == 0` is a bad practice. Here is what I would use:

    "\"length(~ts) == 0\" is discouraged since it has to traverse the whole list to check if it is empty or not. Prefer \"~ts == []\" instead"